File: csv-tools/scripts/diff.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Compare two CSV files: report added, removed, and changed rows (by key column or row order)."""

from __future__ import annotations

import argparse
from typing import Any

from common import load_csv, parse_delimiter, write_json


def add_dialect_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("--delimiter", default=",", help="Field delimiter (use '\\t' for tab).")
    parser.add_argument("--no-header", action="store_true", help="First row is data.")
    parser.add_argument("--skip-lines", type=int, default=None, help="Skip this many preamble lines per file; next line is header.")
    parser.add_argument("--comment-char", help="Skip lines starting with this character.")
    parser.add_argument("--encoding", default="utf-8", help="File encoding.")


def row_key(row: dict[str, str], key_columns: list[str]) -> tuple:
    return tuple((row.get(c, "") or "").strip() for c in key_columns)


def main() -> None:
    parser = argparse.ArgumentParser(description="Diff two CSV files.")
    parser.add_argument("left", help="Left CSV file.")
    parser.add_argument("right", help="Right CSV file.")
    add_dialect_args(parser)
    parser.add_argument("--key", help="Comma-separated key columns for row identity (default: row order).")
    parser.add_argument("--format", choices=["json", "text"], default="json", help="Output format.")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON.")
    args = parser.parse_args()

    dialect = {
        "delimiter": parse_delimiter(args.delimiter),
        "has_header": not args.no_header,
        "comment_char": args.comment_char,
        "encoding": args.encoding,
        "skip_lines": args.skip_lines,
    }

    _, left_rows = load_csv(args.left, **dialect)[:2]
    _, right_rows = load_csv(args.right, **dialect)[:2]

    key_cols: list[str] | None = [c.strip() for c in args.key.split(",")] if args.key else None

    changes: list[dict[str, Any]] = []

    if key_cols:
        left_by_key = {row_key(r, key_cols): r for r in left_rows}
        right_by_key = {row_key(r, key_cols): r for r in right_rows}
        left_keys = set(left_by_key)
        right_keys = set(right_by_key)
        for k in left_keys - right_keys:
            changes.append({"kind": "removed", "key": list(k), "left": left_by_key[k]})
        for k in right_keys - left_keys:
            changes.append({"kind": "added", "key": list(k), "right": right_by_key[k]})
        for k in left_keys & right_keys:
            lr, rr = left_by_key[k], right_by_key[k]
            if lr != rr:
                changes.append({"kind": "changed", "key": list(k), "left": lr, "right": rr})
    else:
        # Row-order diff
        for i, (lr, rr) in enumerate(zip(left_rows, right_rows)):
            if lr != rr:
                changes.append({"kind": "changed", "row_index": i, "left": lr, "right": rr})
        if len(left_rows) > len(right_rows):
            for i in range(len(right_rows), len(left_rows)):
                changes.append({"kind": "removed", "row_index": i, "left": left_rows[i]})
        elif len(right_rows) > len(left_rows):
            for i in range(len(left_rows), len(right_rows)):
                changes.append({"kind": "added", "row_index": i, "right": right_rows[i]})

    if args.format == "text":
        if not changes:
            print("No differences.")
        else:
            for c in changes:
                k = c["kind"]
                if k == "removed":
                    print(f"- removed: {c.get('key', c.get('row_index'))} {c.get('left', {})}")
                elif k == "added":
                    print(f"+ added: {c.get('key', c.get('row_index'))} {c.get('right', {})}")
                else:
                    print(f"~ changed: {c.get('key', c.get('row_index'))}")
        return

    write_json({"change_count": len(changes), "changes": changes}, compact=args.compact)


if __name__ == "__main__":
    main()

File: csv-tools/scripts-node/diff.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import {
  DEFAULT_ENCODING,
  loadCsv,
  parseDelimiter,
  writeJson,
} from "./common.mjs";

function rowKey(row, keyColumns) {
  return keyColumns.map((c) => (row[c] ?? "").trim());
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      delimiter: { type: "string", default: "," },
      "no-header": { type: "boolean", default: false },
      "skip-lines": { type: "string" },
      "comment-char": { type: "string" },
      encoding: { type: "string", default: DEFAULT_ENCODING },
      key: { type: "string" },
      format: { type: "string", default: "json" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  const leftPath = positionals[0];
  const rightPath = positionals[1];
  if (!leftPath || !rightPath) {
    console.error("diff.mjs: requires left and right file paths");
    process.exit(1);
  }

  const dialect = {
    delimiter: parseDelimiter(values.delimiter),
    hasHeader: !values["no-header"],
    commentChar: values["comment-char"] ?? null,
    encoding: values.encoding,
    skipLines: values["skip-lines"] != null ? Number.parseInt(values["skip-lines"], 10) : null,
  };

  const { rows: leftRows } = loadCsv(leftPath, dialect);
  const { rows: rightRows } = loadCsv(rightPath, dialect);

  const keyCols = values.key ? values.key.split(",").map((c) => c.trim()) : null;
  const changes = [];

  if (keyCols && keyCols.length > 0) {
    const leftByKey = new Map();
    const rightByKey = new Map();
    for (const r of leftRows) leftByKey.set(JSON.stringify(rowKey(r, keyCols)), r);
    for (const r of rightRows) rightByKey.set(JSON.stringify(rowKey(r, keyCols)), r);

    const leftKeys = new Set(leftByKey.keys());
    const rightKeys = new Set(rightByKey.keys());

    for (const k of leftKeys) {
      if (!rightKeys.has(k)) {
        changes.push({ kind: "removed", key: JSON.parse(k), left: leftByKey.get(k) });
      }
    }
    for (const k of rightKeys) {
      if (!leftKeys.has(k)) {
        changes.push({ kind: "added", key: JSON.parse(k), right: rightByKey.get(k) });
      }
    }
    for (const k of leftKeys) {
      if (rightKeys.has(k)) {
        const lr = leftByKey.get(k);
        const rr = rightByKey.get(k);
        const lStr = JSON.stringify(lr);
        const rStr = JSON.stringify(rr);
        if (lStr !== rStr) {
          changes.push({ kind: "changed", key: JSON.parse(k), left: lr, right: rr });
        }
      }
    }
  } else {
    const maxLen = Math.max(leftRows.length, rightRows.length);
    for (let i = 0; i < maxLen; i++) {
      const lr = leftRows[i];
      const rr = rightRows[i];
      if (lr === undefined) {
        changes.push({ kind: "added", row_index: i, right: rr });
      } else if (rr === undefined) {
        changes.push({ kind: "removed", row_index: i, left: lr });
      } else if (JSON.stringify(lr) !== JSON.stringify(rr)) {
        changes.push({ kind: "changed", row_index: i, left: lr, right: rr });
      }
    }
  }

  if (values.format === "text") {
    if (changes.length === 0) {
      console.log("No differences.");
    } else {
      for (const c of changes) {
        const k = c.kind;
        const id = c.key ?? c.row_index;
        if (k === "removed") console.log(`- removed: ${JSON.stringify(id)} ${JSON.stringify(c.left ?? {})}`);
        else if (k === "added") console.log(`+ added: ${JSON.stringify(id)} ${JSON.stringify(c.right ?? {})}`);
        else console.log(`~ changed: ${JSON.stringify(id)}`);
      }
    }
    return;
  }

  writeJson({ change_count: changes.length, changes }, values.compact);
}

main();
